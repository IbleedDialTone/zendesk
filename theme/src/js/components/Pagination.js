import preact from 'preact';
import PropTypes from 'prop-types';

import times from 'lodash.times';

const PAGE_BUFFER = 5;

class PaginationItem extends preact.Component {
	constructor(props) {
		super(props);

		this.handleClick = this.handleClick.bind(this);
	}

	handleClick(event) {
		const {onClick} = this.props;

		onClick(event.currentTarget.value);
	}

	render({items}) {
		return (
			<ul>
				{items.map(item => (
					<li class={`pagination-${item.position}`}>
						{item.position !== 'current' && (
							<button class="btn-unstyled" onClick={this.handleClick} type="button" value={item.value}>
								{item.value}
							</button>
						)}

						{item.position === 'current' && (
							<span>{item.value}</span>
						)}
					</li>
				))}
			</ul>
		);
	}
}

PaginationItem.PropTypes = {
	items: PropTypes.shape(
		{
			position: PropTypes.oneOfType(
				[
					PropTypes.number,
					PropTypes.oneOf(
						['current', 'first', 'last', 'next', 'prev']
					)
				]
			),
			url: PropTypes.string,
			value: PropTypes.oneOfType(
				[PropTypes.number, PropTypes.string]
			).isRequired
		}
	),
	onClick: PropTypes.func.isRequired
};

class Pagination extends preact.Component {
	constructor(props) {
		super(props);

		this.handleClick = this.handleClick.bind(this);
		this.setBuffers = this.setBuffers.bind(this);
		this.setPages = this.setPages.bind(this);

		this.state = {
			currentPage: 1
		};
	}

	handleClick(page) {
		this.setState(
			{
				currentPage: parseInt(page)
			}
		)
	}

	setBuffers() {
		const {total} = this.props;
		const {currentPage} = this.state;

		let nextPageBuffer = PAGE_BUFFER;
		let prevPageBuffer = 0;

		if (currentPage > 1) {
			prevPageBuffer =
				currentPage > PAGE_BUFFER ? PAGE_BUFFER : currentPage - 1;

			if (currentPage === total) {
				nextPageBuffer = 0;
			} else {
				nextPageBuffer =
					total - currentPage >= PAGE_BUFFER
						? PAGE_BUFFER
						: total - currentPage;
			}
		}

		return {prevPageBuffer, nextPageBuffer};
	}

	setPages() {
		const {currentPage} = this.state;

		const {prevPageBuffer, nextPageBuffer} = this.setBuffers();

		const startPage = currentPage - prevPageBuffer;
		const totalPages = nextPageBuffer + prevPageBuffer + 1;

		const pages = times(totalPages, i => {
			const value = startPage + i;

			return {
				position: value === currentPage ? 'current' : value,
				url: '/',
				value: value
			};
		});

		if (prevPageBuffer) {
			pages.unshift(
				{
					position: 'first',
					url: '/',
					value: '«'
				},
				{
					position: 'prev',
					url: '/',
					value: '‹'
				}
			);
		}

		if (nextPageBuffer) {
			pages.push(
				{
					position: 'next',
					url: '/',
					value: '›'
				},
				{
					position: 'last',
					url: '/',
					value: '»'
				}
			);
		}

		return pages;
	}

	render() {
		return (
			<nav class="pagination">
				<PaginationItem items={this.setPages()} onClick={this.handleClick} />
			</nav>
		);
	}
}

Pagination.PropTypes = {
	onClick: PropTypes.func.isRequired,
	perPage: PropTypes.number.isRequired,
	total: PropTypes.number.isRequired
};

export default Pagination;