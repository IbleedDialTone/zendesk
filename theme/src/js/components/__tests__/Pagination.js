import preact from 'preact';
import render from 'preact-render-to-string';

import Pagination from '../Pagination';

describe('Pagination', () => {
	it('renders correctly', () => {
		const handleOnclick = () => console.log('clicked');

		const tree = render(
			<Pagination onClick={handleOnclick} perPage={10} total={100} />
		);

		expect(tree).toMatchSnapshot();
	});
});