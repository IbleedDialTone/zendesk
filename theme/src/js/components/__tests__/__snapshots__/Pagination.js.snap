// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<nav class="pagination">
  <ul>
    <li class="pagination-current">
      <span>1</span>
    </li>
    <li class>
      <button class="btn-unstyled" type="button" value="2">2</button>
    </li>
    <li class>
      <button class="btn-unstyled" type="button" value="3">3</button>
    </li>
    <li class>
      <button class="btn-unstyled" type="button" value="4">4</button>
    </li>
    <li class>
      <button class="btn-unstyled" type="button" value="5">5</button>
    </li>
    <li class>
      <button class="btn-unstyled" type="button" value="6">6</button>
    </li>
    <li class>
      <button class="btn-unstyled" type="button" value="2">›</button>
    </li>
    <li class>
      <button class="btn-unstyled" type="button" value="100">»</button>
    </li>
  </ul>
</nav>
`;
